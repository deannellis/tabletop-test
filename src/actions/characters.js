import uuid from 'uuid';

// ADD_CHARACTER
export const addCharacter = (
    {
        abilities = {
            strength: 0,
            intelligence: 0,
            wisdom: 0,
            dexterity: 0,
            constitution: 0,
            charisma: 0
        }, 
        race = '', 
        charClass = '', 
        hp = 0, 
        xp = 0,
        name = '',
        description = ''
    } = {}
) => ({
    type: 'ADD_CHARACTER',
    character: {
        id: uuid(),
        abilities,
        race,
        charClass,
        hp,
        xp,
        name,
        description
    }
});

// REMOVE_CHARACTER

export const removeCharacter = ({ id } = {}) => ({
    type: 'REMOVE_CHARACTER',
    id
});

// EDIT_CHARACTER
export const editCharacter = (id, updates) => ({
    type: 'EDIT_CHARACTER',
    id,
    updates
});

