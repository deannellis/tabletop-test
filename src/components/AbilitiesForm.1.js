import React from 'react';

export default class AbilitiesForm extends React.Component {
    state = {
        strength: undefined,
        intelligence: undefined,
        wisdom: undefined,
        dexterity: undefined,
        constitution: undefined,
        charisma: undefined
    };
    
    renderButtons = () => {
        Object.keys(this.state).map((key) => {
            return 
            <div>
                <button type="button" onClick={this.onRollStrength} disabled={typeof this.state.strength != 'undefined'}>Strength</button>
                <p>Strength: {typeof this.state.strength == 'undefined' ? 'Roll to determine Strength' : this.state.strength }</p>
            
        });
    };
    
    onRollStrength = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({strength: rollValue}));
    };
    
    onRollIntelligence = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({intelligence: rollValue}));
    };
    
    onRollWisdom = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({wisdom: rollValue}));
    };
    
    onRollDexterity = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({dexterity: rollValue}));
    };
    
    onRollConstitution = () => {
        let rollValue = Math.floor(Math.random() * ((20 + 1) - 1)) + 1;
        this.setState(() => ({constitution: rollValue}));
    };
    
    render() {
        return (
            <div>
                <form>
                    {this.renderButtons()}
                </form>
            </div>
        );
    }
}