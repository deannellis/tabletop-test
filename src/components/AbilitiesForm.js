import React from 'react';
import SingleRollDie from './SingleRollDie';
import uuid from 'uuid';

export default class AbilitiesForm extends React.Component {
    constructor(props){
        super(props);
        
        this.state = {
            strength: [],
            intelligence: [],
            wisdom: [],
            dexterity: [],
            constitution: [],
            charisma: [],
            error: '',
            refreshKey: 0,
            lastDieRolled: undefined,
            lastResult: undefined
        };
    }
    
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit} >
                    {this.renderAttribute('Strength', 'strength')}
                    {this.renderAttribute('Intelligence', 'intelligence')}
                    {this.renderAttribute('Wisdom', 'wisdom')}
                    {this.renderAttribute('Dexterity', 'dexterity')}
                    {this.renderAttribute('Constitution', 'constitution')}
                    {this.renderAttribute('Charisma', 'charisma')}
                    <button>Next Step</button>
                </form>
            </div>
        );
    }
    
    handleRollDie = (result, attributeKey) => {
        this.setState(() => ({
            lastResult: result,
            [attributeKey]: [...this.state[attributeKey], result]
        }));
    };
    
    onSubmit = (e) => {
        e.preventDefault();
        
        const sum = (accumulator, currentValue) => accumulator + currentValue;
        
        if(this.state.strength.length < 3 || this.state.intelligence.length < 3 || this.state.wisdom.length < 3 || this.state.dexterity.length < 3 || this.state.constitution.length < 3 || this.state.charisma.length < 3) {
            this.setState(() => ({
                error: 'Please roll all scores'
            }));
        } else {
            this.setState(() => ({
                error: ''
            }));
            this.props.onSubmit({
                id: uuid(),
                abilities: {
                    strength: this.state.strength.reduce(sum),
                    intelligence: this.state.intelligence.reduce(sum),
                    wisdom: this.state.wisdom.reduce(sum),
                    dexterity: this.state.dexterity.reduce(sum),
                    constitution: this.state.constitution.reduce(sum),
                    charisma: this.state.charisma.reduce(sum)
                },
                name: 'In-Progress Character',
                charClass: 'unkown'
            });
        } 
    }
    
    renderAttribute = (label, attributeKey) => {
        const attributeValue = this.state[attributeKey];
        const isDefined = attributeValue.length !== 0;
        let rollsString = '';
        if(this.state[attributeKey].length < 3){
            this.state[attributeKey].forEach(
                (roll) => {rollsString += `${roll} + `}
            );
        } else {
            let total = 0;
            this.state[attributeKey].forEach(
                (roll) => {total += roll}
            );
            rollsString = `${this.state[attributeKey][0]} + ${this.state[attributeKey][1]} + ${this.state[attributeKey][2]} = ${total}`;
        }
        
        const message = label + ': ' + (isDefined ? rollsString : 'roll to determine');
        return(
            <div>
                <p>{message}</p>
                <div className="dice-container">
                    <SingleRollDie dieType={6} handleRollDie={this.handleRollDie} attributeKey={attributeKey} updateRefreshKey={this.updateRefreshKey} refreshKey={this.state.refreshKey}/>
                    <SingleRollDie dieType={6} handleRollDie={this.handleRollDie} attributeKey={attributeKey} updateRefreshKey={this.updateRefreshKey} refreshKey={this.state.refreshKey}/>
                    <SingleRollDie dieType={6} handleRollDie={this.handleRollDie} attributeKey={attributeKey} updateRefreshKey={this.updateRefreshKey} refreshKey={this.state.refreshKey}/>
                </div>
            </div>
        );
    }
    
    updateRefreshKey = (newKey) => {
        this.setState(() => ({
            refreshKey: newKey
        }));
    };
}