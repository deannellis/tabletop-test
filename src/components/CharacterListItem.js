import React from 'react';
import { Link } from 'react-router-dom';

const CharacterListItem = ({ name, charClass, race, id }) => (
    <div>
        <h2>{name}</h2>
        <p>{race} - {charClass}</p>
        <Link to={`/view-char/${id}`}>View Character</Link>
    </div>    
);

export default CharacterListItem;
