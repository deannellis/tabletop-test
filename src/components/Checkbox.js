import React, { Component, PropTypes } from 'react';

class Checkbox extends React.Component {
    state = {
        isChecked: false,
    }
    
    toggleCheckboxChange = () => {
        const { handleCheckboxChange } = this.props;
        const { isChecked } = this.state;
    
        this.setState(({ isChecked }) => (
            {
                isChecked: !isChecked,
            }
        ));
    
        handleCheckboxChange(!isChecked);
    }
    
      render() {
        const { label } = this.props;
        const { isChecked } = this.state;
        let isDisabled;
        if(isChecked) {
            isDisabled = false;
        } else {
            isDisabled = this.props.isDisabled
        }
    
        return (
          <div className="checkbox">
            <label>
                <input
                    type="checkbox"
                    value={isChecked}
                    checked={isChecked}
                    onChange={this.toggleCheckboxChange}
                    disabled={isDisabled}
                />
    
              {label}
            </label>
          </div>
        );
      }
}

export default Checkbox;