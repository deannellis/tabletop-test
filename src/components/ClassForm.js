import React from 'react';
import { connect } from 'react-redux';
import spells from '../objects/spells';
import SpellListItem from './SpellListItem';

class ClassForm extends React.Component {
    constructor(props){
        super(props);
        
        this.state = {
            selectedOption: '',
            error: ''
        };
    }
    
    render() {
        const characters = this.props.characters;
        const lastCharacter = characters[characters.length - 1];
        let canCleric = true;
        let canFighter = true;
        let canMagicUser = true;
        let canThief = true;
        
        if(lastCharacter.race == 'dwarf') {
            canMagicUser = false;
        } else if (lastCharacter.race == 'halfling') {
            canFighter = false;
        } 
        
        if(lastCharacter.abilities.wisdom < 9) {
            canCleric = false;
        }
        
        if(lastCharacter.abilities.strength < 9) {
            canFighter = false;
        }
        
        if(lastCharacter.abilities.intelligence < 9) {
            canMagicUser = false;
        }
        
        if(lastCharacter.abilities.dexterity < 9) {
            canThief = false;
        }
        
        return (
            <div>
                <h2>Select Your Class</h2>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit} >
                    {canCleric 
                        ? 
                        <div>
                            <label>
                                <input type="radio" name="class" value="cleric"
                                    checked={this.state.selectedOption === 'cleric'}
                                    onChange={this.handleOptionChange}
                                />
                                Cleric
                            </label>
                        </div> 
                        : 
                        <p>Cleric: Your ability scores are insufficient</p>
                    }
                    {canFighter 
                        ? 
                        <div>
                            <label>
                                <input type="radio" name="class" value="fighter"
                                    checked={this.state.selectedOption === 'fighter'}
                                    onChange={this.handleOptionChange}
                                />
                                Fighter
                            </label>
                        </div> 
                        : 
                        <p>Fighter: Your ability scores are insufficient</p>
                    }
                    {canMagicUser 
                        ? 
                        <div>
                            <label>
                                <input type="radio" name="class" value="magic-user"
                                    checked={this.state.selectedOption === 'magic-user'}
                                    onChange={this.handleOptionChange}
                                />
                                Magic-User
                            </label>
                        </div> 
                        : 
                        <p>Magic-User: Your ability scores are insufficient</p>
                    }
                    {canThief 
                        ? 
                        <div>
                            <label>
                                <input type="radio" name="class" value="thief"
                                    checked={this.state.selectedOption === 'thief'}
                                    onChange={this.handleOptionChange}
                                />
                                Thief
                            </label>
                        </div> 
                        : 
                        <p>Thief: Your ability scores are insufficient</p>
                    }
                    <button>Next Step</button>
                </form>
            </div>
        )
    }
    
    handleOptionChange = (e) => {
        this.setState({
            selectedOption: e.target.value
        });
    }
    
    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.selectedOption) {
            this.setState(() => ({
                error: 'Please select a class'
            }));
        } else {
            const characters = this.props.characters;
            const lastCharacter = characters[characters.length - 1];
            this.setState(() => ({
                error: ''
            }));
            this.props.onSubmit( lastCharacter.id, {
                charClass: this.state.selectedOption
            });
        }
    }
}

const mapStateToProps = (state) => {
    return {
        characters: state.characters
    };
};

export default connect(mapStateToProps)(ClassForm);