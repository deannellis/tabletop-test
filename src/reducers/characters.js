//  CHARACTERS REDUCER
import actionIds from 'actions/actionIds';
const {ADD_CHARACTER, REMOVE_CHARACTER, EDIT_CHARACTER, COMPLETE_CHARACTER} = actionIds;
const charactersReducerDefaultState = [];

export default ( state = charactersReducerDefaultState, action ) => {
    switch (action.type) {
        case ADD_CHARACTER:
            return [
                ...state,
                action.character
            ];
        case REMOVE_CHARACTER:
            return state.filter(({ id }) => {
                return id !== action.id;
            });
        case EDIT_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    return {
                        ...character,
                        ...action.updates
                    }
                } else {
                    return character;
                }
            });
        case COMPLETE_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    let {inProgressStep, ...rest} = character;
                    console.log('rest'+rest);
                    return rest;
                } else {
                    return character;
                }
            });
        default:
            return state;
    }
};