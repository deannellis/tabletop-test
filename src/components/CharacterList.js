import React from 'react';
import { connect } from 'react-redux';
import CharacterListItem from './CharacterListItem';

const CharacterList = (props) => (
    <div>
        <h1>Character List</h1>
        {props.characters.length > 0 ?
            <div>
                {props.characters.map((character) => {
                    return <CharacterListItem key={character.id} {...character} />
                })}
            </div>
        :
            <p>You have no characters, Click below to create a character</p>
        }
    </div>    
);

const mapStateToProps = (state) => {
    return {
        characters: state.characters
    };
};

export default connect(mapStateToProps)(CharacterList);
