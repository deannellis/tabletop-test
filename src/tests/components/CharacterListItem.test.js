/* global expect */
import React from 'react';
import { shallow } from 'enzyme';
import { CharacterListItem } from '../../components/CharacterListItem';
import characters from '../fixtures/characters';

test('should render CharacterListItem with a character that has inProgressStep key', () => {
    const wrapper = shallow(<CharacterListItem {...characters[0]} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render CharacterListItem with a character that does not have inProgressStep key', () => {
    let {inProgressStep, ...rest} = characters[0];
    const wrapper = shallow(<CharacterListItem {...rest} />);
    expect(wrapper).toMatchSnapshot();
});