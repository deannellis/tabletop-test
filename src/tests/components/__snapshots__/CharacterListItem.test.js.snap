// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CharacterListItem with a character that does not have inProgressStep key 1`] = `
<div>
  <h2>
    Rick Slick
  </h2>
  <p>
    human
     
     - magic-user
  </p>
  <Link
    replace={false}
    to="/view-char/3a7b71ef-fb4e-4bde-a2c8-17cd5b9f4097"
  >
    View Character
  </Link>
</div>
`;

exports[`should render CharacterListItem with a character that has inProgressStep key 1`] = `
<div>
  <h2>
    Rick Slick
  </h2>
  <p>
    human
     
     - magic-user
  </p>
  <Link
    replace={false}
    to="/new-char-step-2/3a7b71ef-fb4e-4bde-a2c8-17cd5b9f4097"
  >
    Finish Character
  </Link>
</div>
`;
